gradle/lucene-dev/lucene-dev-repo-composite.gradle

```diff
@@ -0,0 +1,159 @@
+import java.util.function.BiFunction
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Local Lucene development repository resolution:
+//   1) A "-Plucene.dev.version=[version]" property, resolving Lucene artifacts from a local Maven repository.
+//   2) A non-empty property "-Plucene.dev.path=[path]" pointing to a local path.
+//   3) An auto-wired 'lucene' subfolder, if present. To skip auto-wiring, pass
+//      a blank value in step 2: "-Plucene.dev.path=".
+
+
+// This script is applied in settings.gradle and later at build time: these two contexts
+// are distinctively different and have separate (and limited) APIs.
+def configuringSettings = (rootProject instanceof org.gradle.api.initialization.ProjectDescriptor)
+
+// Accessor for -P properties from settings or at build time.
+BiFunction<String, String, String> resolvePropertyValue = { propertyName, defValue ->
+  if (configuringSettings) {
+    return settings.startParameter.projectProperties.getOrDefault(propertyName, defValue)
+  } else {
+    return project.properties.getOrDefault(propertyName, defValue)
+  }
+}
+
+def PROP_FORCE_VERSION="lucene.dev.version"
+def PROP_FORCE_PATH="lucene.dev.path"
+def READ_ACCESS_PROPERTY="lucene-dev-path.dir"
+
+
+def forcedLuceneVersion = resolvePropertyValue(PROP_FORCE_VERSION, null)
+if (forcedLuceneVersion != null) {
+  if (!configuringSettings) {
+    logger.lifecycle("Lucene version forced by -P${PROP_FORCE_VERSION}=${forcedLuceneVersion}")
+
+    allprojects {
+      repositories {
+        mavenLocal()
+      }
+
+      tasks.withType(Test) {
+        def userHome = System.properties.get('user.home')
+        systemProperty READ_ACCESS_PROPERTY, file("${userHome}/.m2/repository/org/apache/lucene").absolutePath
+      }
+
+      configurations.all {
+        resolutionStrategy.eachDependency {
+          if (requested.group == "org.apache.lucene") {
+            useVersion(forcedLuceneVersion)
+            because("Lucene version forced manually by 'lucene.dev.version' property.")
+          }
+        }
+      }
+    }
+  }
+
+  // Step 1: end resolution
```

Comment:
I've tested this out with various combinations, and clearing all the prerelease5 stuff out of `~/.gradle/cache/` and mavenLocal, etc., and everything works as expected. The only thing I'm wondering about is whether it's worth being a little more verbose in logging for the case where `lucene.dev.version` supersedes one of the other configs; e.g., before returning early after step 1:
```
if (propForcePathValue != null ? !propForcePathValue.isBlank() : defaultLuceneDevRepo.exists()) {
  logger.lifecycle("`lucene.dev.version` supersedes local lucene dev repo")
}
```
... or something like that.

I think it's perfectly fine without the extra log messages though.